### PR TITLE
Implement kernel and userspace symbol lookup for remote targets via BPFd

### PR DESCRIPTION
**Pull request by jcanseco**
This patch:

- Updates the BPFd source in BCC to be in sync with the recent changes related to kernel/user symbol lookup in the BPFd repo
- Fixes a couple of bugs concerning how BCC interacts with BPFd
- Does a bit of clean up and removes some of the HiKey demo hacks
- **Hooks up BCC's kernel/userspace symbol lookup logic to BPFd**, allowing for remote symbol lookup in BCC

### CODE REVIEW

**Comment by jcanseco** [comment]
This was tested with the following tools:
- cachestat.py
- cachetop.py
- offcputime.py

All of which seemed to work just fine.

**Comment by jcanseco** [comment]
I also ran the following tools to test this PR. None of these tools worked, but the problem does not seem to be the result of this patch judging from the error messages and considering the fact that we get the exact same results pre-patch.

**stacksnoop.py** submit_bio 

Fails to compile when run with HiKey as remote:
```
/virtual/main.c:10:19: error: use of undeclared identifier '__START_KERNEL_map'
        if (ret < __START_KERNEL_map)
                  ^
/virtual/main.c:27:14: error: no member named 'ip' in 'struct pt_regs'
    if (ctx->ip)
        ~~~  ^
/virtual/main.c:28:45: error: no member named 'ip' in 'struct pt_regs'
        bpf_trace_printk("ip: %llx\n", ctx->ip);
                                       ~~~  ^
/virtual/main.c:29:15: error: no member named 'bp' in 'struct pt_regs'
    bp = ctx->bp;
         ~~~  ^
4 errors generated.
Traceback (most recent call last):
  File "./tools/old/stacksnoop.py", line 103, in <module>
    b = BPF(text=bpf_text)
  File "/usr/lib/python2.7/dist-packages/bcc/__init__.py", line 317, in __init__
    raise Exception("Failed to compile BPF module %s" % src_file)
Exception: Failed to compile BPF module 
```

Runtime error when run on ShellRemote OR when run on the Host (i.e. with no remotes set):

```
TIME(s)            STACK
Traceback (most recent call last):
  File "./tools/old/stacksnoop.py", line 121, in <module>
    (reg, addr) = msg.split(" ")
ValueError: too many values to unpack
```

I think the reason stacksnoop.py doesn't work with ShellRemote and Host is because this version of stacksnoop may just be too old and so its integration with BCC is outdated. However, the fact that running it on HiKey produces a different error might be a cause of concern.

**stacksnoop.lua** submit_bio 

Fails to compile when run on either HiKey, ShellRemote, or Host:
```
[ERROR] bcc.lua:2251: bcc-lua: undefined symbol: bcc_symcache_new
stack traceback:
	[C]: in function '__index'
	bcc.lua:2251: in function 'SymbolCache'
	./tools/stacksnoop.lua:53: in function <./tools/stacksnoop.lua:44>
	[C]: in function 'xpcall'
	bcc.lua:5894: in function <bcc.lua:5843>
	[C]: at 0x00844640
```
Question: was stacksnoop (either the .py or the .lua version) ever working for you? 

I initially suspected that it might be due to me lacking Lua dependencies on my workstation, but I ran biosnoop.lua as well and that seemed to work fine.

**profile.py**

Runtime exception when run on HiKey or ShellRemote:
```
Sampling at 49 Hertz of all threads by user + kernel stack... Hit Ctrl-C to end.
ioctl(PERF_EVENT_IOC_SET_BPF) failed: Bad file descriptor
Traceback (most recent call last):
  File "./tools/profile.py", line 219, in <module>
    sample_period=0, sample_freq=args.frequency)
  File "/usr/lib/python2.7/dist-packages/bcc/__init__.py", line 860, in attach_perf_event
    sample_period, sample_freq, pid, i, group_fd)
  File "/usr/lib/python2.7/dist-packages/bcc/__init__.py", line 847, in _attach_perf_event
    raise Exception("Failed to attach BPF to perf event")
Exception: Failed to attach BPF to perf event
```
However, works just fine when run on Host.

I created the following GitHub issues to record these findings: [#31](https://github.com/joelagnel/bpfd/issues/31), [#32](https://github.com/joelagnel/bpfd/issues/32)

File: src/python/bcc/remote/libremote.py [hunk 1]
```diff
@@ -249,7 +249,6 @@ def perf_reader_poll(self, fd_callbacks, timeout):
             raw_cb(ct.cast(id(self), ct.py_object), data_bin, ct.c_int(size))
 
     def close_connection(self):
-        self._remote_send_command("exit")
```
Comment (joelagnel) [inline src/python/bcc/remote/libremote.py]: The correct fix should be to make BPFd send the response back. Can you please do that instead?

**Review by joelagnel — CHANGES_REQUESTED** [review]

**Comment by joelagnel** [comment]
You should investigate into why in remote mode the tools don't work. The symbol look ups should work whether on host or remote. I don't see any reason off the top why. As for the tools erroring out, for me they run just fine without the symbol look up code, the only issue is the symbols show "unknown". So I'm reluctant to pull your patches, until that is straightened out / debugged further. Also I feel you should do this investigation / debugging prior to upstreaming the remote patches

**Comment by jcanseco** [comment]
> As for the tools erroring out, for me they run just fine without the symbol look up code, the only issue is the symbols show "unknown".

This sticks out to me since the tools error out for me even _without_ the symbol patches, so the fact that you're not seeing the same errors implies that this might be a problem unique to my set-up. I just want to make sure: stacksnoop.py, stacksnoop.lua, and profile.py run just fine for you without errors regardless if you run on HiKey, ShellRemote, and Host?

What happens if you try running the tools in your set-up after applying the symbol patches (i.e. by cloning a copy of my [fork](github.com/jcanseco/bcc) and checking out the `symbols` branch) ?

**Comment by jcanseco** [comment]
Reminder: we discussed this in person the other day and noted that it was not stacksnoop.py and profile.py that you ran that were working, but rather stackcount.py. We concluded to run stackcount.py as a test for this PR, and to consider the errors for stacksnoop.py and profile.py as issues separate from this PR.

Well I tested this with stackcount.py and it seemed to work fine. I think we should be in a position to approve this PR now unless you have other concerns.

**Comment by jcanseco** [comment]
Note: I also decided to add the smoke test changes to this PR instead of letting it have its own PR (formerly PR #9). It made more sense this way.

**Comment by jcanseco** [comment]
I addressed the requests you made regarding the "exit" command and the HiKey demo hacks.

**Comment by joelagnel** [comment]
Great, thanks!

**Pull request merged**